lib/sync/watcher.go

```diff
@@ -0,0 +1,269 @@
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"boscoin.io/sebak/lib/block"
+	"boscoin.io/sebak/lib/common"
+	"boscoin.io/sebak/lib/errors"
+	"boscoin.io/sebak/lib/network"
+	"boscoin.io/sebak/lib/node"
+	"boscoin.io/sebak/lib/node/runner/api"
+	"boscoin.io/sebak/lib/storage"
+	"github.com/inconshreveable/log15"
+)
+
+type Watcher struct {
+	syncer    SyncController
+	cm        network.ConnectionManager
+	st        *storage.LevelDBBackend
+	localNode *node.LocalNode
+	client    Doer
+	after     AfterFunc
+	interval  time.Duration
+	stop      chan chan struct{}
+	ctx       context.Context
+	cancel    context.CancelFunc
+	logger    log15.Logger
+}
+
+func NewWatcher(
+	syncer SyncController,
+	client Doer,
+	cm network.ConnectionManager,
+	st *storage.LevelDBBackend,
+	ln *node.LocalNode) *Watcher {
+	ctx, cancel := context.WithCancel(context.Background())
+	w := &Watcher{
+		syncer:    syncer,
+		client:    client,
+		cm:        cm,
+		st:        st,
+		localNode: ln,
+		after:     time.After,
+		interval:  5 * time.Second,
+		ctx:       ctx,
+		cancel:    cancel,
+		logger:    common.NopLogger(),
+	}
+	return w
+}
+
+func (w *Watcher) SetLogger(l log15.Logger) {
+	w.logger = l
+}
+
+func (w *Watcher) Start() error {
+	w.localNode.SetWatch()
+	w.loop()
+	return nil
+}
+
+func (w *Watcher) Stop() error {
+	w.cancel()
+	c := make(chan struct{})
+	w.stop <- c
+	<-c
+	return nil
+}
+
+func (w *Watcher) loop() {
+	var (
+		syncer        = w.syncer
+		checkc        = w.after(w.interval)
+		highestHeight uint64
+		latestHeight  uint64
+		nodes         []string
+		err           error
+		ctx           = w.ctx
+	)
+
+	latestHeight = w.latestHeight()
+	w.logger.Info("starting sync watcher", "height", latestHeight)
+
+L:
+	for {
+		select {
+		case <-checkc:
+			highestHeight, nodes, err = w.highestHeightAndNodes(ctx)
+			if err != nil {
+				if err == context.Canceled {
+					break L
+				}
+				w.logger.Error("get highest height has err", "err", err, "high", highestHeight, "nodes", nodes)
+				checkc = w.after(w.interval)
+				continue
+			}
+			if highestHeight > latestHeight {
+				w.logger.Info("set sync target block", "high", highestHeight, "nodes", nodes)
+				syncer.SetSyncTargetBlock(ctx, highestHeight, nodes)
+				latestHeight = highestHeight
+			}
+			w.logger.Info("watched sync height", "high", highestHeight, "last", latestHeight)
+			checkc = w.after(w.interval)
+		case c := <-w.stop:
+			close(c)
+			break L
+		}
+	}
+	w.logger.Info("end sync watcher", "high", highestHeight, "last", latestHeight)
+
+}
+
+func (w *Watcher) highestHeightAndNodes(ctx context.Context) (uint64, []string, error) {
+	select {
+	case <-ctx.Done():
+		return 0, nil, ctx.Err()
+	default:
+	}
+
+	nodes, err := w.fetchNodeInfos(ctx)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	height, err := w.bestHeightFromNodes(ctx, nodes)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	nodeAddrs, err := w.bestNodeAddrs(ctx, height, nodes)
```

Comment:
a node of nodes may have a lower height than the height derived from `w.bestHeightFromNodes(ctx, nodes)`
It means that syncer may fail with this call `syncer.SetSyncTargetBlock(ctx, highestHeight, nodes)`
I am not sure how the `SetSyncTargetBlock` method works. Could you consider this?

Reply:
Syncer try to fetch this best height, but if Syncer can't fetch this best height then Syncer retry from current height to best height until fetch to success. 
the best height is from node 's state is 'CONSENSUS`, It means all nodes is going to go this best height.